finat/enriched.py

```diff
@@ -118,3 +129,24 @@ def tree_map(f, *args):
         return tuple(tree_map(f, *subargs) for subargs in zip(*args))
     else:
         return f(*args)
+
+
+def concatenate_entity_dofs(ref_el, elements, method):
+    """Combine the entity DoFs from a list of elements into a combined
+    dict containing the information for the concatenated DoFs of all
+    the elements.
+
+    :arg ref_el: the reference cell
+    :arg elements: subelement whose DoFs are concatenated
+    :arg method: method to obtain the entity DoFs dict
+    :returns: concatenated entity DoFs dict
+    """
+    entity_dofs = {dim: {i: [] for i in entities}
+                   for dim, entities in iteritems(ref_el.get_topology())}
+    offsets = numpy.cumsum([0] + list(e.space_dimension()
+                                      for e in elements), dtype=int)
+    for i, d in enumerate(map(method, elements)):
+        for dim, dofs in iteritems(d):
+            for ent, off in iteritems(dofs):
+                entity_dofs[dim][ent] += list(map(partial(add, offsets[i]), off))
```

Comment:
These come out sorted because the sub entities are sorted, and enriched doesn't interleave.